chore: 🤖 adds deploy script

ozwaldorf opened the pull request:
### Why?

Using icx heavily overcomplicates the scripts and requires users to build an extra tool.
It also did not work on the public ic mainnet.

### Demo

`./healthcheck.sh`

`./deploy.sh local`

ozwaldorf committed:
Better healthcheck and dedicated deploy script, that works on any network. dfx.json updated to use latest rust container type

punkbit commented on deploy.sh:
```diff
@@ -0,0 +1,74 @@
+
+if [[ -z $1 ]];
+then
+    printf "DEPLOY.SH:\n\n   usage: ./deploy.sh <local|ic|other> [reinstall]\n\n"
```
First of all, thanks for taking the time to share your proposal.

Mind adding the command as a script in the package.json, as we have for the others? Also, add to the documentation `develop/docs/nft-example.md`, how to work with it, a basic description is enough.

punkbit commented on dfx.json:
```diff
@@ -2,12 +2,13 @@
-      "wasm": "target/wasm32-unknown-unknown/release/nft.wasm",
-      "type": "custom"
-    }
+    "canisters": {
+      "nft": {
+    	"package": "nft",
+    	"candid": "nft/candid/nft.did",
+    	"type": "rust"
+      }
+    },
```
Mind fix the following, as we have nested "canisters" for field "canisters"?

```suggestion
      "nft": {
    	"package": "nft",
    	"candid": "nft/candid/nft.did",
    	"type": "rust"
      },
```

Ref:
https://smartcontracts.org/docs/release-notes/0.8.4-rn.html

punkbit reviewed (changes requested):
Thanks for your time, the effort is appreciated!

Left a few comments below, which request some amends, adding the scripts to the package.json as we are using it in the project to keep it simple, and also need to add to the documentation, in this case for the example nft implementation, please.

Afterwards, we'll have to run a few tests and tweak a little bit, if that's ok. Thank you!

ozwaldorf committed:
chore: 🤖 Added new scripts to package.json

chore: 🤖 Moved scripts back to `.scripts/`

chore: 🤖 `deploy.sh` now checks if process has been run previously

fix: 🐛 Remove nested `canister` item in dfx.json

ozwaldorf committed:
chore: 🤖 Added new scripts to package.json

chore: 🤖 Moved scripts back to `.scripts/`

chore: 🤖 `deploy.sh` now checks if process has been run previously

fix: 🐛 Remove nested `canister` item in dfx.json

ozwaldorf committed:
docs: ✏️ Updated nft example docs to use the new scripts

ozwaldorf committed:
docs: ✏️ Updated nft example docs to use the new scripts

heldrida commented on package.json:
```diff
@@ -9,8 +9,8 @@
 		"cap:start": "cd ./cap && dfx deploy ic-history-router && mkdir -p ../.temp && dfx canister id ic-history-router > ../.temp/ic-history-router-id",
-		"dip721:healthcheck": "./.scripts/healthcheck.sh",
-		"dip721:deploy": "./.scripts/deploy.sh",
+		"healthcheck": "./.scripts/healthcheck.sh",
+		"deploy": "./.scripts/deploy.sh",
```
Given the original convention, let's keep the prefix for the service, as we have for the Cap dependency, please.

```
"dip721:healthcheck": "./.scripts/healthcheck.sh",
"dip721:deploy": "./.scripts/deploy.sh",
```

heldrida commented on docs/nft-example.md:
````diff
@@ -26,11 +26,13 @@ You have to pull the `CAP` submodule content as follows:
 
 ```sh
 yarn cap:init
+
+npm run cap:init
````
That's alright, but the documentation already mentions that'll be using "yarn" as an example and the user can use npm or any other package manager if much prefer. So, to avoid having to show every single command for any other package manager, we accept that yarn is the base of the example, so no need to show the npm example.

"💡 During the guide we'll be using `yarn`, but this can be easily replaced by `npm`, if that's your preference"

```suggestion

```

punkbit replied:
The documentation already mentions that'll be using "yarn" as an example and the user can use npm or any other package manager if much prefer. So, to avoid having to show every single command for any other package manager, we accept that yarn is the base of the example, so no need to show the npm example.

```suggestion
```

punkbit commented on docs/nft-example.md:
````diff
@@ -42,6 +44,8 @@ Once ready, launch the healtcheck for our Nft implementation example by running
 
 ```sh
 yarn healthcheck
+
+npm run healthcheck
````
```suggestion
```

punkbit commented on docs/nft-example.md:
````diff
@@ -50,16 +54,34 @@ Through the process the [Cap Service](https://github.com/Psychedelic/cap) is dep
 
 ```sh
 yarn cap:start
+
+npm run cap:start
````
```suggestion
```

punkbit commented on docs/nft-example.md:
````diff
@@ -50,16 +54,34 @@ Through the process the [Cap Service](https://github.com/Psychedelic/cap) is dep
+
+```sh
+yarn deploy <local|ic> [reinstall]
+
+npm run deploy <local|ic> [reinstall]
````
```suggestion
```

punkbit commented on docs/nft-example.md:
````diff
@@ -50,16 +54,34 @@ Through the process the [Cap Service](https://github.com/Psychedelic/cap) is dep
+npm run deploy <local|ic> [reinstall]
+```
+
+💡 On creation of a new canister, the controllers have to be set to the principalid of the owner and the canister id. this script automatically sets that up, and redeploys or reinstalls if the canister already exists
````
```suggestion
💡 On the creation of a new canister, the controllers have to be set to the owners principal and the canister id, as required by the ownership and approvals handling - the script automatically sets that up; And can also redeploy or reinstalls, if the canister already exists.
```

punkbit commented on docs/nft-example.md:
````diff
@@ -50,16 +54,34 @@ Through the process the [Cap Service](https://github.com/Psychedelic/cap) is dep
 
 ```sh
 yarn dev:watch
+
+npm start dev:watch
````
```suggestion
```

punkbit commented on package.json:
```diff
@@ -1,17 +1,18 @@
+	"scripts": {
+		"cap:init": "git submodule update --init --recursive",
+		"cap:start": "cd ./cap && dfx deploy ic-history-router && mkdir -p ../.temp && dfx canister id ic-history-router > ../.temp/ic-history-router-id",
+		"healthcheck": "./.scripts/healthcheck.sh",
+		"deploy": "./.scripts/deploy.sh",
```
Let's follow the original convention for the services and have the name as a prefix, as we do for the Cap service.

```suggestion
		"dip721:healthcheck": "./.scripts/healthcheck.sh",
		"dip721:deploy": "./.scripts/deploy.sh",
```

punkbit reviewed (changes requested):
Thanks for taking the time and effort, left a small review and some requests for changes 👍

ozwaldorf committed:
docs: ✏️ Remove unnecessary npm commands in nft examples
chore: 🤖 Fix deploy and healhcheck to follow dip721:script convention

ozwaldorf commented:
You're welcome, I look forward to working with you guys in the future!

ozwaldorf committed:
chore: 🤖 healthcheck.sh now uses `npm run dip721:deploy`

ozwaldorf commented:
Were there any other change requests for me to complete?

punkbit commented:
> Were there any other change requests for me to complete?

Hi, there's nothing to worry about for now, appreciate your patience, you'll get notified in the next update. Thanks 👍

ozwaldorf committed:
fix: 🐛 Canister controller not being correctly set in deploy.sh

punkbit commented:
⚠️ Included in the integration of https://github.com/Psychedelic/DIP721/pull/31
